Zend/zend_compile.c

```diff
@@ -3719,31 +3723,45 @@ void zend_compile_global_var(zend_ast *ast) /* {{{ */
 }
 /* }}} */
 
+static uint32_t zend_lookup_static_var(zend_op_array *op_array, zend_string *name) /* {{{ */ {
+	uint32_t i = 0;
+	zend_ulong hash_value = zend_string_hash_val(name);
+
+	while (i < op_array->last_static_var) {
+		if (ZSTR_VAL(op_array->static_vars[i]) == ZSTR_VAL(name) ||
+		    (ZSTR_H(op_array->static_vars[i]) == hash_value &&
+		     ZSTR_LEN(op_array->static_vars[i]) == ZSTR_LEN(name) &&
+		     memcmp(ZSTR_VAL(op_array->static_vars[i]), ZSTR_VAL(name), ZSTR_LEN(name)) == 0)) {
+			zend_string_release(name);
+			return i;
+		}
+		i++;
+	}
+	i = op_array->last_static_var++;
+	if (op_array->last_static_var > CG(context).static_vars_size) {
+		CG(context).static_vars_size += 8;
+		op_array->static_vars = erealloc(op_array->static_vars, CG(context).static_vars_size * sizeof(zend_string*));
+		op_array->static_variables = erealloc(op_array->static_variables, CG(context).static_vars_size * sizeof(zval));
+	}
+	op_array->static_vars[i] = zend_new_interned_string(name);
+	return i;
+}
+/* }}} */
+
 static void zend_compile_static_var_common(zend_ast *var_ast, zval *value, zend_bool by_ref) /* {{{ */
 {
 	znode var_node, result;
 	zend_op *opline;
+	uint32_t var_num;
 
 	zend_compile_expr(&var_node, var_ast);
 
-	if (!CG(active_op_array)->static_variables) {
-		if (CG(active_op_array)->scope) {
-			CG(active_op_array)->scope->ce_flags |= ZEND_HAS_STATIC_IN_METHODS;
-		}
-		ALLOC_HASHTABLE(CG(active_op_array)->static_variables);
-		zend_hash_init(CG(active_op_array)->static_variables, 8, NULL, ZVAL_PTR_DTOR, 0);
-	}
-
-	if (GC_REFCOUNT(CG(active_op_array)->static_variables) > 1) {
-		if (!(GC_FLAGS(CG(active_op_array)->static_variables) & IS_ARRAY_IMMUTABLE)) {
-			GC_REFCOUNT(CG(active_op_array)->static_variables)--;
-		}
-		CG(active_op_array)->static_variables = zend_array_dup(CG(active_op_array)->static_variables);
-	}
-	zend_hash_update(CG(active_op_array)->static_variables, Z_STR(var_node.u.constant), value);
+	var_num = zend_lookup_static_var(CG(active_op_array), Z_STR(var_node.u.constant));
+	ZVAL_COPY_VALUE(&CG(active_op_array)->static_variables[var_num], value);
 
-	opline = zend_emit_op(&result, by_ref ? ZEND_FETCH_W : ZEND_FETCH_R, &var_node, NULL);
-	opline->extended_value = ZEND_FETCH_STATIC;
+	opline = zend_emit_op(&result, ZEND_FETCH_LEXICAL, NULL, NULL);
+	opline->op1.num = var_num;
+	opline->extended_value = by_ref? ZEND_FETCH_W : ZEND_FETCH_R;
```

Comment:
Better `FETCH_STATIC`. It's not necessarily lexical.

Also, should we combine this  with the following assign op into `BIND_STATIC`, similar to what we do for globals?